### PR TITLE
Refactor Project to Gradle

### PR DESCRIPTION
**Pull request by JRoy**
Gradle is better than Maven, don't @ me. okay but actually it's [faster](https://www.youtube.com/watch?v=atuFSv2bLa8&feature=youtu.be&t=77), compiles and tests in parallel more efficiently, and more epic stuff).

Testing:
- [x] Builds
- [x] Dependencies are shaded correctly
- [x] Plugins run without error
- [x] Checkstyle rules added
- [x] Verify that Crowdin works correctly (@mdcfe)
- [x] New Maven Repo (@mdcfe)

### CODE REVIEW

File: shadow.gradle [hunk 1]
```diff
@@ -0,0 +1,14 @@
+apply plugin: "com.github.johnrengelman.shadow"
+
+configurations {
+    shade
```
Comment (zml2008) [inline shadow.gradle]: why not `shadow`? this isn't maven.

File: settings.gradle [hunk 2]
```diff
@@ -0,0 +1,36 @@
+rootProject.name = 'EssentialsXParent'
+
+include(':EssentialsX')
```
Comment (zml2008) [inline settings.gradle]: You can reduce duplication here by doing something like:

```groovy

[
  "",
  "AntiBuild",
  "Chat",
  /* ... */
].each {
  include(":EssentialsX$it")
  project(":EssentialsX$it").projectDir = file("Essentials$it")
}
```

File: settings.gradle [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+project(':EssentialsXProtect').projectDir = file('EssentialsProtect')
+
+include(':EssentialsXSpawn')
+project(':EssentialsXSpawn').projectDir = file('EssentialsSpawn')
+
+include(':EssentialsXXMPP')
+project(':EssentialsXXMPP').projectDir = file('EssentialsXMPP')
+
+// Providers
+include(':BaseProviders')
```
Comment (zml2008) [inline settings.gradle]: maybe skip the custom project directory, just do `include(":providers:BaseProviders")` and update the project deps to match?

File: build.gradle [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+tasks.shadowJar.dependsOn tasks.relocateShadowJar
+
+subprojects {
+    apply plugin: 'java'
+    apply plugin: 'java-library'
+    apply plugin: 'checkstyle'
+
+    checkstyle {
```
Comment (zml2008) [inline build.gradle]: You should be able to just set the config directory here: https://github.com/KyoriPowered/indra/blob/master/indra-common/src/main/kotlin/net/kyori/indra/IndraCheckstylePlugin.kt#L42

File: build.gradle [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        maven { url = 'https://hub.spigotmc.org/nexus/content/groups/public/' }
+        maven { url = 'https://papermc.io/repo/repository/maven-public/' }
+        maven { url = 'https://jitpack.io' }
+    }
+
+    dependencies {
+        testImplementation 'junit:junit:4.12'
+        testImplementation 'org.mockito:mockito-core:3.2.0'
+
+        if (project.name != "1_8Provider" && project.name != "PaperProvider" && project.name != "NMSReflectionProvider") { // These providers use their own bukkit versions
```
Comment (zml2008) [inline build.gradle:70]: You should be able to do an `afterEvaluate` block, where you only add `spigot-api` as a dependency if no version of spigot or its forks has been added as a dep yet. That'd let you do this without hardcoding specific project names.

File: build.gradle [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            api "org.spigotmc:spigot-api:${spigotVersion}"
+        }
+    }
+
+    buildscript {
```
Comment (zml2008) [inline build.gradle]: You shouldn't be using this -- the `plugins` block at the top is preferred.

If you want a plugin that isn't applied to the root project but is available to subprojects, just add `apply false` after the version.

File: build.gradle [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    processResources {
+        filter(ReplaceTokens, beginToken: '${',
+                endToken: '}', tokens: ["full.version": "${version}.${GIT_DEPTH}".toString()])
+    }
+
+    sourceCompatibility = '1.8'
+    compileJava.options.encoding = 'UTF-8'
+
+    tasks.withType(JavaCompile) {
+        options.encoding = 'UTF-8'
```
Comment (zml2008) [inline build.gradle]: Apply the `release` flag here: https://github.com/KyoriPowered/indra/blob/master/indra-common/src/main/kotlin/net/kyori/indra/IndraPlugin.kt#L57

File: Essentials/src/main/java/com/earth2me/essentials/utils/StringUtil.java [hunk 8]
```diff
@@ -8,6 +8,7 @@
 public final class StringUtil {
     private static final Pattern INVALIDFILECHARS = Pattern.compile("[^a-z0-9-]");
     private static final Pattern STRICTINVALIDCHARS = Pattern.compile("[^a-z0-9]");
+    @SuppressWarnings("CheckStyle")
```
Comment (zml2008) [inline Essentials/src/main/java/com/earth2me/essentials/utils/StringUtil.java:11]: If you're trying to suppress something from Checkstyle, you should be suppressing individual checks, not everything. https://checkstyle.org/config_filters.html#SuppressWarningsFilter might be useful

File: build.gradle [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+    tasks.build.dependsOn tasks.checkstyleMain
+
+    repositories {
+        mavenLocal()
+        mavenCentral()
+        maven { url = 'https://ci.ender.zone/plugin/repository/everything/' }
+        maven { url = 'https://hub.spigotmc.org/nexus/content/groups/public/' }
+        maven { url = 'https://papermc.io/repo/repository/maven-public/' }
+        maven { url = 'https://jitpack.io' }
```
Comment (zml2008) [inline build.gradle]: Jitpack should probably be specifically restricted to artifacts that supposed come from there -- at the moment is that just Vault?

https://docs.gradle.org/current/userguide/declaring_repositories.html#sec:repository-content-filtering

File: build.gradle [hunk 10]
```diff
@@ -0,0 +1,143 @@
+plugins {
+    id 'org.ajoberstar.grgit' version '4.1.0'
+    id 'com.github.johnrengelman.shadow' version '6.1.0'
+    id 'java'
```
Comment (zml2008) [inline build.gradle]: Why are you applying the `java` and `shadow` plugins to the root project? There isn't any Java source here.

File: build.gradle [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+def outputTasks() {
+    [":EssentialsX:shadowJar", ":EssentialsXAntiBuild:jar", ":EssentialsXChat:jar",
+     ":EssentialsXGeoIP:shadowJar", ":EssentialsXProtect:jar", ":EssentialsXSpawn:shadowJar",
+     ":EssentialsXXMPP:shadowJar"].stream().map({ tasks.findByPath(it) })
+}
+
+task copyToJars(type: Copy) {
+    outputTasks().forEach {
+        println it
+        from(it.outputs.files)
```
Comment (zml2008) [inline build.gradle]: you should be able to just do `from(it)` here -- then you don't have to explicitly declare this task's dependency on the output tasks.

File: build.gradle [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+build.dependsOn copyToJars
+clean.dependsOn cleanJars
+copyToJars.dependsOn tasks.findByPath(":EssentialsX:processResources")
+tasks.build.dependsOn tasks.shadowJar
+
+outputTasks().forEach { copyToJars.dependsOn it }
+
+compileJava.options.encoding = 'UTF-8'
```
Comment (zml2008) [inline build.gradle]: I assume the stuff from here on is left over from an earlier attempt?

File: gradle/wrapper/gradle-wrapper.properties [hunk 13]
```diff
@@ -0,0 +1,6 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.6.1-bin.zip
+zipStoreBase=GRADLE_USER_HOME
+zipStorePath=wrapper/dists
+org.gral
```
Comment (zml2008) [inline gradle/wrapper/gradle-wrapper.properties]: is this supposed to be here?

File: .gitignore [hunk 14]
```diff
@@ -17,6 +17,6 @@
 /Essentials/usermap.csv
 
 # Build files
-target/
+.gradle/
 jars/
-out/
```
Comment (zml2008) [inline .gitignore:22]: `out/` shows up from some IDE setups, probably shouldn't remove it

**Review by zml2008 — CHANGES_REQUESTED** [review]
hooray!

You might also want to take a look at [spigradle](https://github.com/spigradle/spigradle) -- it has some neat features for Bukkit plugin development.

File: shadow.gradle (hunk 1 above)
Comment (JRoy) [reply shadow.gradle]: shadow defines the configuration `shadow` and freaks out if you try to steal it from it

File: gradle/wrapper/gradle-wrapper.properties (hunk 13 above)
Comment (JRoy) [reply gradle/wrapper/gradle-wrapper.properties]: oops

File: build.gradle (hunk 12 above)
Comment (JRoy) [reply build.gradle]: yeah :P

File: build.gradle (hunk 11 above)
Comment (JRoy) [reply build.gradle]: yep! also spot the missed debug println 😄

File: shadow.gradle (hunk 1 above)
Comment (zml2008) [reply shadow.gradle]: it defines the configuration -- you should be able to just go ahead and reuse that shadow configuration when configuring the shadow tasks.

File: Essentials/src/main/java/com/earth2me/essentials/utils/StringUtil.java (hunk 8 above)
Comment (JRoy) [reply Essentials/src/main/java/com/earth2me/essentials/utils/StringUtil.java:11]: I don't want to fully suppress it, it's just this one place

File: Essentials/build.gradle [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    compileOnly 'net.luckperms:api:5.0'
+
+    api 'io.papermc:paperlib:1.0.6-SNAPSHOT'
+
+    // Providers
+    api project(':providers:BaseProviders')
+    api project(':providers:PaperProvider')
+    api project(':providers:NMSReflectionProvider')
+    api(project(':providers:1_8Provider')) {
+        exclude group: 'bukkit'
```
Comment (mdcfe) [inline Essentials/build.gradle]: `group: bukkit`?

File: EssentialsSpawn/build.gradle [hunk 16]
```diff
@@ -0,0 +1,10 @@
+dependencies {
+    implementation project(':EssentialsX')
```
Comment (mdcfe) [inline EssentialsSpawn/build.gradle]: Shouldn't this be `compileOnly`?

File: EssentialsSpawn/build.gradle [hunk 17]
```diff
@@ -0,0 +1,10 @@
...
+}
+
+shadowJar {
+    dependencies {
+        include (dependency('io.papermc:paperlib'))
+    }
+    relocate 'io.papermc.lib', 'com.earth2me.essentials.spawn.paperlib'
```
Comment (mdcfe) [inline EssentialsSpawn/build.gradle]: We could consider using EssentialsX's shaded PaperLib instead of shading our own

File: build.gradle [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+subprojects {
+    apply plugin: 'net.kyori.indra'
+    apply plugin: 'net.kyori.indra.checkstyle'
+    apply plugin: 'net.kyori.indra.publishing'
+    apply plugin: 'com.github.johnrengelman.shadow'
+
+    repositories {
+        mavenLocal()
```
Comment (mdcfe) [inline build.gradle:54]: Is this necessary?

File: build.gradle [hunk 19]
```diff
@@ -0,0 +1,161 @@
+plugins {
+    id 'org.ajoberstar.grgit' version '4.1.0'
+    id 'net.kyori.indra' version '1.0.2' apply false
+    id 'net.kyori.indra.checkstyle' version '1.0.2' apply false
+    id 'net.kyori.indra.publishing' version '1.0.2' apply false
+    id 'com.github.johnrengelman.shadow' version '6.1.0' apply false
+    id 'java'
```
Comment (mdcfe) [inline build.gradle]: Todo: make root not depend on java plugin, if possible

File: build.gradle [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+subprojects {
+    apply plugin: 'net.kyori.indra'
+    apply plugin: 'net.kyori.indra.checkstyle'
+    apply plugin: 'net.kyori.indra.publishing'
+    apply plugin: 'com.github.johnrengelman.shadow'
+
+    repositories {
+        mavenLocal()
+        mavenCentral()
+        maven { url = 'https://ci.ender.zone/plugin/repository/everything/' }
```
Comment (mdcfe) [inline build.gradle]: Also do we still need this?

File: build.gradle [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+                }
+                ciManagement {
+                    system = 'Jenkins'
+                    url = 'https://ci.ender.zone/job/EssentialsX'
+                }
+            }
+        }
+    }
+
+    tasks.withType(Sign) {
```
Comment (mdcfe) [inline build.gradle]: KyoriPowered/indra#20

File: build.gradle [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+    processResources {
+        filter(ReplaceTokens, beginToken: '${',
+                endToken: '}', tokens: ["full.version": fullVersion])
+    }
+
+    indra {
+        checkstyle = "$checkstyleVersion"
+
+        github('EssentialsX', 'Essentials')
+        license = new License(
```
Comment (mdcfe) [inline build.gradle]: KyoriPowered/indra#19

File: providers/1_8Provider/build.gradle [hunk 23]
```diff
@@ -0,0 +1,5 @@
+dependencies {
+    api(project(':providers:NMSReflectionProvider')) {
+        exclude group: 'bukkit'
```
Comment (mdcfe) [inline providers/1_8Provider/build.gradle]: Again `group: bukkit`?

File: providers/PaperProvider/build.gradle [hunk 24]
```diff
@@ -0,0 +1,4 @@
+dependencies {
+    implementation project(':providers:BaseProviders')
+    compileOnly 'com.destroystokyo.paper:paper-api:1.16.3-R0.1-SNAPSHOT'
```
Comment (mdcfe) [inline providers/PaperProvider/build.gradle:3]: 1.16.4?

Comment (JRoy) [reply providers/PaperProvider/build.gradle:3]: Doesn't need to be updated until we need a new method for paper provider

File: .github/workflows/build.yml [hunk 25]
```diff
@@ -2,37 +2,47 @@ name: Build EssentialsX
...
-      run: mvn package verify --file pom.xml
-    - name: Copy artifacts
-      run: mkdir -p ./out/ && cp -t ./out/ **/target/Essentials*.jar
-    - uses: actions/upload-artifact@master
+    - name: Build with Gradle
+      run: |
+        chmod +x gradlew
+        ./gradlew build --stacktrace
+    - name: Deploy with Gradle
+      if: ${{ success() && github.event_name == 'push' && github.ref == 'refs/head/2.x' }}
```
Comment (mdcfe) [inline .github/workflows/build.yml]:
```suggestion
      if: ${{ success() && github.event_name == 'push' && github.repository == 'EssentialsX/Essentials' && github.ref == 'refs/head/2.x' }}
```

**Review by mdcfe — APPROVED** [review]

**Pull request merged**